pkg/block/mdfetcher.go

```diff
@@ -0,0 +1,103 @@
+package block
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+)
+
+type MetadataFetcher struct {
+	cachedBlockMDs map[ulid.ULID]*metadata.Meta
+	mu             *sync.RWMutex
+
+	concurrencyLimit int
+	logger           log.Logger
+	bkt              objstore.BucketReader
+}
+
+func NewMetadataFetcher(logger log.Logger, concurrencyLimit int, bkt objstore.BucketReader) *MetadataFetcher {
+	return &MetadataFetcher{
+		logger:           logger,
+		concurrencyLimit: concurrencyLimit,
+		mu:               &sync.RWMutex{},
+		cachedBlockMDs:   make(map[ulid.ULID]*metadata.Meta),
+		bkt:              bkt,
+	}
+}
+
+// Fetch wil return the mapping from block ID to its metadata that are currently stored in the bucket passed in.
+// If the corresponding value for a block ID key is nil, no valid metadata is retrieved successfully.
+func (f *MetadataFetcher) Fetch(ctx context.Context) (map[ulid.ULID]*metadata.Meta, error) {
+	blockIDs := make(chan ulid.ULID, f.concurrencyLimit)
+	defer close(blockIDs)
+	remoteBlockMDs := make(map[ulid.ULID]*metadata.Meta, len(f.cachedBlockMDs))
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < f.concurrencyLimit; i++ {
+		go func() {
+			for blockID := range blockIDs {
+				md, err := f.fetchBlockMDIfNotCached(ctx, wg, blockID)
+				if err != nil {
+					level.Debug(f.logger).Log("msg", "failed to fetch block metadata", "block", blockID, "error", err)
+				}
+
+				f.mu.Lock()
+				remoteBlockMDs[blockID] = md
```

Comment:
I think there is a race currently: we call `wg.Done` before inserting the returned values in the map, meaning `wg.Done()` may return early and we may return an incomplete map to the caller.